runtime/shared_common/OSCache.cpp

```diff
@@ -739,20 +741,56 @@ SH_OSCache::getAllCacheStatistics(J9JavaVM* vm, const char* ctrlDirName, UDATA g
 
 	/* Walk all the cache files in the j9shmem dir and (optionally) the persistent cacheDir */
 	do {
-		if (getCacheStatistics(vm, ctrlDirName, (const char*)nameWithVGen, groupPerm, localVerboseFlags, j2seVersion, &tempInfo, reason) != -1) {
+		J9Pool* lowerLayerList = NULL;
+		J9Pool** lowerLayerListPrt = &lowerLayerList;
+		bool isToplayer = isTopLayerCache(vm, ctrlDirName, nameWithVGen);
+		bool getLowerLayerStats = false;
+
+		if (getCacheStatistics(vm, ctrlDirName, (const char*)nameWithVGen, groupPerm, localVerboseFlags, j2seVersion, &tempInfo, reason, getLowerLayerStats, isToplayer, lowerLayerListPrt, NULL) != -1) {
 			if (includeOldGenerations || !tempInfo.isCompatible || (getCurrentCacheGen() == tempInfo.generation)) {
 				if (tempInfo.isCompatible) {
 					if (!ignoreCompatible) {
-						newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
-						memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+						if (isToplayer) {
+							/* nameWithVGen is a compatible top layer cache, lower layer stats are in lowerLayerList*/
+							newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+							memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+							if (NULL != lowerLayerList) {
+								if (0 < pool_numElements(lowerLayerList)) {
+									pool_state poolState;
+									memset(&poolState, 0, sizeof(pool_state));
+									SH_OSCache_Info* anElement = (SH_OSCache_Info*)pool_startDo(lowerLayerList, &poolState);
+									do {
+										newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+										if (NULL == newElement) {
+											break;
+										}
+										memcpy(newElement, anElement, sizeof(SH_OSCache_Info));
+										++cntr;
+									} while ((anElement = (SH_OSCache_Info*)pool_nextDo(&poolState)) != NULL);
+								}
+								pool_kill(lowerLayerList);
```

Comment:
There needs to be a pool_kill(lowerLayerList) even when getCacheStatistics() fails, as it could have been allocated. Although I think it should be done in SH_CacheMap::startupForStats().